stdlib/public/core/SequenceAlgorithms.swift

```diff
@@ -707,15 +707,29 @@ extension Sequence {
   /// - Complexity: O(*n*), where *n* is the length of the sequence.
   @inlinable
   public __consuming func reversed() -> [Element] {
-    // FIXME(performance): optimize to 1 pass?  But Array(self) can be
-    // optimized to a memcpy() sometimes.  Those cases are usually collections,
-    // though.
-    var result = Array(self)
-    let count = result.count
-    for i in 0..<count/2 {
-      result.swapAt(i, count - ((i + 1) as Int))
+    let underestimatedCount = self.underestimatedCount
+    if underestimatedCount == 0 {
+      var result = Array(self)
+      let count = result.count
+      for i in 0..<count/2 {
+        result.swapAt(i, count - ((i + 1) as Int))      }
+      return result
+    } else {
+      var iterator = makeIterator()
+      var result = [Element](unsafeUninitializedCapacity: underestimatedCount) { buf, initedCount in
+        for i in 0..<underestimatedCount {
+          guard let next = iterator.next() else {
+              preconditionFailure("underestimatedCount greater than count")
+          }
+          buf[underestimatedCount - i - 1] = next
+        }
+        initedCount = underestimatedCount
+      }
+      while let next = iterator.next() {
+        result.insert(next, at: 0)
```

Comment:
This will turn sequences that provide a non-zero underestimate into a quadratic operation for the undercounted elements. This might be rare, but I don't think we can do that. If it is rare, a reasonable compromise may be to revert to building the array in full at this point and then reversing it.

Reply:
You are right. I will fix this.